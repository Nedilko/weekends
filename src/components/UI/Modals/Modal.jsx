import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import Button from '@UI/Buttons/Button'
import Panel from '@UI/Panels/Panel'

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired,
  onApply: PropTypes.func.isRequired,
  onCancel: PropTypes.func,
  isButtonsCentered: PropTypes.bool,
  isHeadingCentered: PropTypes.bool,
}

Modal.defaultProps = {
  isButtonsCentered: false,
  isHeadingCentered: false,
}

function Modal({
  children,
  title,
  onApply,
  onCancel,
  isButtonsCentered,
  isHeadingCentered,
}) {
  return ReactDOM.createPortal(
    <div className="relative z-10">
      <div
        className="fixed inset-0 flex items-center justify-center bg-black/30 backdrop-blur-sm"
        onClick={onCancel}
      >
        <Panel size="xl">
          <div
            className={`mb-2 text-3xl font-thin text-slate-600 dark:border-b-zinc-600 dark:text-gray-400 ${
              isHeadingCentered ? 'text-center' : ''
            }`}
          >
            {title}
          </div>
          {children}
          <div
            className={`flex dark:border-t-zinc-600 ${
              isButtonsCentered ? 'justify-center' : 'justify-end'
            }`}
          >
            <Button onClick={onApply} title="Apply" />
            {onCancel && <Button onClick={onCancel} title="Cancel" />}
          </div>
        </Panel>
      </div>
    </div>,
    document.getElementById('modal-root')
  )
}

export default Modal
