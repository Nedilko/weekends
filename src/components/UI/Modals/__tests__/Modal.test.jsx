import Modal from '@UI/Modals/Modal'
import { render, screen, userEvent } from '@utils/test-utils'

describe('Modal', () => {
  const applyHandler = jest.fn()
  const cancelHandler = jest.fn()

  beforeEach(() => {
    applyHandler.mockClear()
    cancelHandler.mockClear()
  })

  it('should render modal content', () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    render(
      <Modal title="sample title" onApply={applyHandler}>
        <div>sample content</div>
      </Modal>
    )
    expect(screen.getByText('sample content')).toBeInTheDocument()
    document.body.removeChild(modalRoot)
  })

  it('should has title', () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    render(
      <Modal title="sample title" onApply={applyHandler}>
        <div>sample content</div>
      </Modal>
    )
    expect(screen.getByText('sample title')).toBeInTheDocument()
    document.body.removeChild(modalRoot)
  })

  it('should handle Cancel button click', async () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const user = userEvent.setup()
    render(
      <Modal
        title="sample title"
        onApply={applyHandler}
        onCancel={cancelHandler}
      >
        <div>sample content</div>
      </Modal>
    )
    await user.click(screen.getByRole('button', { name: 'Cancel' }))
    expect(cancelHandler).toHaveBeenCalledTimes(1)
    document.body.removeChild(modalRoot)
  })

  it('should handle Apply button click', async () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const user = userEvent.setup()
    render(
      <Modal title="sample title" onApply={applyHandler}>
        <div>sample content</div>
      </Modal>
    )
    await user.click(screen.getByRole('button', { name: 'Apply' }))
    expect(applyHandler).toHaveBeenCalledTimes(1)
    document.body.removeChild(modalRoot)
  })

  it('should center buttons', () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const { container } = render(
      <Modal
        title="sample title"
        onApply={applyHandler}
        isButtonsCentered={true}
      >
        <div>sample content</div>
      </Modal>
    )
    expect(container).toMatchSnapshot()
    document.body.removeChild(modalRoot)
  })

  it('should center heading', () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const { container } = render(
      <Modal
        title="sample title"
        onApply={applyHandler}
        isHeadingCentered={true}
      >
        <div>sample content</div>
      </Modal>
    )
    expect(container).toMatchSnapshot()
    document.body.removeChild(modalRoot)
  })
})
