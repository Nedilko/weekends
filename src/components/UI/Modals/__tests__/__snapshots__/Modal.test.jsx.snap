// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should center buttons 1`] = `
<div
  id="modal-root"
>
  <div>
    <div
      class="relative z-10"
    >
      <div
        class="fixed inset-0 flex items-center justify-center bg-black/30 backdrop-blur-sm"
      >
        <div
          class="max-w-xl mx-auto rounded-xl bg-white p-4 shadow-md dark:bg-zinc-900"
        >
          <div
            class="mb-2 text-3xl font-thin text-slate-600 dark:border-b-zinc-600 dark:text-gray-400 "
          >
            sample title
          </div>
          <div>
            sample content
          </div>
          <div
            class="flex dark:border-t-zinc-600 justify-center"
          >
            <button
              class="mx-4 w-16 py-2 text-xl font-thin text-slate-600 duration-200 ease-in-out hover:scale-110 focus:scale-110 focus:outline-none dark:text-gray-400"
            >
              Apply
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should center heading 1`] = `
<div
  id="modal-root"
>
  <div>
    <div
      class="relative z-10"
    >
      <div
        class="fixed inset-0 flex items-center justify-center bg-black/30 backdrop-blur-sm"
      >
        <div
          class="max-w-xl mx-auto rounded-xl bg-white p-4 shadow-md dark:bg-zinc-900"
        >
          <div
            class="mb-2 text-3xl font-thin text-slate-600 dark:border-b-zinc-600 dark:text-gray-400 text-center"
          >
            sample title
          </div>
          <div>
            sample content
          </div>
          <div
            class="flex dark:border-t-zinc-600 justify-end"
          >
            <button
              class="mx-4 w-16 py-2 text-xl font-thin text-slate-600 duration-200 ease-in-out hover:scale-110 focus:scale-110 focus:outline-none dark:text-gray-400"
            >
              Apply
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
